lib/sync/context.ts

```diff
@@ -0,0 +1,534 @@
+	discourse: [
+		{
+			local: 'page-',
+			remote: '_page',
+		},
+	],
+};
+export interface Logger {
+	warn: (message: string, data: any) => void;
+	error: (message: string, data: any) => void;
+	debug: (message: string, data: any) => void;
+	info: (message: string, data: any) => void;
+	exception: (message: string, error: any) => void;
+}
+export interface SyncActionContext {
+	logger: Logger;
+	getLocalUsername: (username: string) => string;
+	getRemoteUsername: (username: string) => string;
+	getElementBySlug: (
+		slug: string,
+		usePrivilegedSession?: boolean,
+	) => Promise<Contract | null>;
+	getElementById: (id: string) => Promise<Contract | null>;
+	getElementByMirrorId: (
+		type: string,
+		mirrorId: string,
+		options: { usePattern?: boolean },
+	) => Promise<Contract | null>;
+	upsertElement: (
+		type: string,
+		object:
+			| Partial<Contract>
+			| { id: string; type: string; patch: Operation[] },
+		options: { actor?: string; timestamp?: Date; originator?: string },
+	) => Promise<Contract | null>;
+}
+
+export interface IntegrationExecutionContext {
+	logger: Logger;
+	getLocalUsername: (username: string) => string;
+	getRemoteUsername: (username: string) => string;
+	getElementBySlug: (
+		slug: string,
+		usePrivilegedSession?: boolean,
+	) => Promise<Contract | null>;
+	getElementById: (id: string) => Promise<Contract | null>;
+	getElementByMirrorId: (
+		type: string,
+		mirrorId: string,
+		options: { usePattern?: boolean },
+	) => Promise<Contract | null>;
+	request: <T>(
+		actorId: string,
+		requestOptions: any,
+	) => Promise<{ code: number; body: T }>;
+	getActorId: (information: ActorInformation) => Promise<string>;
+}
+
+/**
+ * @name getSyncActionContext
+ * @description This function generates a "context" object that provides a common interface
+ *  for use by sync integrations.
+ * @param {String} provider - The name of the integration e.g. 'github', 'discourse'
+ * @param {Object} workerContext - Context object provided to action functions by
+ * the Jellyfish worker
+ * @param {Object} context - Logging context object
+ * @param {String} session - session token used to interact with Jellyfish
+ *
+ * @returns {Object}
+ *
+ * @example
+ *
+ * const handler = async (session, context, card, request) => {
+ * 	const syncContext = context.sync.getActionContext(request.arguments.provider,
+ * 			context, request.context, context.privilegedSession)
+ * 	)
+ * 	.....
+ * }
+ */
+export const getSyncActionContext = (
+	workerContext: WorkerContext,
+	provider: string,
+	context: any,
+	session: string,
+): SyncActionContext => {
+	const getDefaultActor = async (): Promise<null | string> => {
+		const sessionCard = await workerContext.getCardById(session, session);
+		if (!sessionCard) {
+			return null;
+		}
+		// TODO: Replace this return type with the session contract interface
+		return sessionCard.data.actor as string;
+	};
+
+	const contextObject: SyncActionContext = {
+		logger: {
```

Comment:
IMO this makes more sense either with these functions as part of the root level of the object, or with `logger` renamed to `log` so:

```
context.log.error(...)
```

instead of

```
context.logger.error(...)
```

Reply:
Hmm, won't that make naming inconsistent? AFAIK logging methods are derived from a `Logger` interface obtained through a `getLogger` method. What's the motivation for naming it `log` here or for making these functions part of the root level object?